### PR TITLE
plugin/kubernetes: Add upstream @self and loop count

### PR DESCRIPTION
**Pull request by chrisohaver**
### 1. What does this pull request do?

- Adds `@self` option to upstream for kubernetes and etcd plugins.  `upstream @self` will cause upstream `CNAME` target lookups to be processed by the current server (via recursive re-entry into dnsserver.ServeDNS()).
- Counts re-entries into ServeDNS(), and breaks loop if re-entries > limit

Still To Do:
- ~~Document `@self` option~~
- ~~Integration tests~~

### 2. Which issues (if any) are related?

coredns/deployment#55
#1427

### 3. Which documentation changes (if any) need to be made?

Document `@self` option in `upstream` directive in k8s and etcd plugin READMEs.

### CODE REVIEW

**Comment by miekg** [comment]
We should drop the `@self` so we don't introduce a new namespace.

File: plugin/backend.go
Comment (miekg) [inline plugin/backend.go]: Should we fumble context.Context in state? `opt` makes sense as well, but context.Context is not really an option.
opt is also nice because of the symmetry with `Reverse`

File: core/dnsserver/server.go
Comment (miekg) [inline core/dnsserver/server.go]: docs?

File: core/dnsserver/server.go
Comment (miekg) [inline core/dnsserver/server.go]: What about tcp?

File: core/dnsserver/server.go
Comment (miekg) [inline core/dnsserver/server.go]: `return ctx, nil` and close the if

File: core/dnsserver/server.go
Comment (miekg) [inline core/dnsserver/server.go]: must it be exported?
the full var now reads `dnsserver.ServerKey` think just `Key` read better?

File: core/dnsserver/server.go
Comment (miekg) [inline core/dnsserver/server.go]: incrementDepthAndCheck (or something better - depthCheck doesn't convey it is also incrementing)

File: plugin/backend.go
Comment (johnbelamaric) [reply plugin/backend.go]: fumble? What are you trying to suggest?

Comment (miekg) [reply plugin/backend.go]: should request.Request get a context.Context member in the struct.

Comment (johnbelamaric) [reply plugin/backend.go]: Oh, you mean "bundle" :)

Yeah, that could work. `opt` already existed in some places which is why Chris used it, I don't know if that place uses request.Request or not - @chrisohaver ?

Comment (johnbelamaric) [reply plugin/backend.go]: Yeah, in all these https://github.com/coredns/coredns/pull/1484/files#diff-347a95a2d36b7b336cb55cc146bd16d6R36 - they have `state` too, so that would work and be less change. But what's the point of Options in all those...

File: core/dnsserver/server.go
Comment (johnbelamaric) [reply core/dnsserver/server.go]: I have seen others use a typed, empty struct as the key. That way there is no collision with someone else using the context. See https://github.com/opentracing/opentracing-go/blob/master/gocontext.go#L12

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP: Add upst..." ]
>johnbelamaric commented on this pull request.
>
>
>
>> @@ -18,7 +19,7 @@ type ServiceBackend interface {
> 	Reverse(state request.Request, exact bool, opt Options) ([]msg.Service, error)
>
> 	// Lookup is used to find records else where.
>-	Lookup(state request.Request, name string, typ uint16) (*dns.Msg, error)
>+	Lookup(state request.Request, name string, typ uint16, opt Options) (*dns.Msg, error)
>
>Yeah, in all these https://github.com/coredns/coredns/pull/1484/files#diff-347a95a2d36b7b336cb55cc146bd16d6R36 - they have `state` too, so that would work and be less change. But what's the point of Options in all those...

force_tcp is one of those options. There could be more in the future, having an 
Options struct would make adding those easier and does not break the API.

File: plugin/backend.go
Comment (chrisohaver) [reply plugin/backend.go]: yeah - its not really an "option."  I'll add it to request.Request.

**Comment by chrisohaver** [comment]
I seem to have botched up my fork's master when testing the docker release.

I'll try to restore it.

**Comment by chrisohaver** [comment]
Testing new upstream integration tests coredns/ci#42: /integration-cipr42

**Comment by chrisohaver** [comment]
Integration tests added in coredns/ci#42, will merge that once this merges.

File: request/request.go
Comment (johnbelamaric) [inline request/request.go]: We use the external context package everywhere else and should use it here too.

File: plugin/backend.go
Comment (johnbelamaric) [inline plugin/backend.go]: can be on one line, right?

**Review by johnbelamaric — CHANGES_REQUESTED** [review]

**Review by johnbelamaric — APPROVED** [review]

**Comment by johnbelamaric** [comment]
@miekg PTAL, we need to merge this soon as K8s deadlines are coming up (we'll need 1.0.6 this week)

**Comment by johnbelamaric** [comment]
FYI, I believe we added `context` to the `state` for this. We may want to reconsider this at some point, given the explicit go recommendation NOT to do this. See https://golang.org/pkg/context/

We would have to make a breaking change to put Context as the first arg in all these calls. Maybe a 1.1 thing?

**Comment by miekg** [comment]
Hmmm..good point. Let's revert some commits then...

**Comment by johnbelamaric** [comment]
No, I think we should go with this for 1.0.6 and then modify it in 1.1. We can't fix the problem without a breaking change, otherwise. And we need it fixed for k8s.

**Comment by miekg** [comment]
Ok. Clear.

**Pull request merged**
